viz/metrics-api/edges.go

```diff
@@ -26,48 +25,77 @@ var formatMsg = map[string]string{
 	"not_provided_by_service_discovery": "Not Provided By Service Discovery",
 }
 
+type edgeKey struct {
+	src   string
+	srcNs string
+	dst   string
+	dstNs string
+}
+
 func (s *grpcServer) Edges(ctx context.Context, req *pb.EdgesRequest) (*pb.EdgesResponse, error) {
 	log.Debugf("Edges request: %+v", req)
 	if req.GetSelector().GetResource() == nil {
 		return edgesError(req, "Edges request missing Selector Resource"), nil
 	}
 
-	edgesHTTP, err := s.getEdges(ctx, req, "response_total")
-	if err != nil {
-		return edgesError(req, err.Error()), nil
-	}
-	edgesTCP, err := s.getEdges(ctx, req, "tcp_open_total")
+	resourceType := promResourceType(req.GetSelector().GetResource())
+	dstResourceType := "dst_" + resourceType
+	labelsOutbound := promDirectionLabels("outbound")
+	labelsOutboundStr := generateLabelStringWithExclusion(labelsOutbound, string(resourceType), string(dstResourceType))
+	query := fmt.Sprintf(edgesQuery, "tcp_open_connections", labelsOutboundStr, resourceType, resourceType)
+
+	promResult, err := s.queryProm(ctx, query)
 	if err != nil {
 		return edgesError(req, err.Error()), nil
 	}
 
-	edges := []*pb.Edge{}
-	// iterate over tcp_open_total metrics
-	for _, edgeTCP := range edgesTCP {
-		edgeHTTPIndex := -1
-		// find the corresponding response_total entry
-		for i, edgeHTTP := range edgesHTTP {
-			if equalEdges(edgeTCP, edgeHTTP) {
-				edgeHTTPIndex = i
-				break
+	edgeMap := make(map[edgeKey]*pb.Edge)
+
+	for _, sample := range promResult {
+		if sample.Value == 0.0 {
+			continue
+		}
+		key := edgeKey{
+			src:   string(sample.Metric[resourceType]),
+			srcNs: string(sample.Metric[model.LabelName("namespace")]),
+			dst:   string(sample.Metric[dstResourceType]),
+			dstNs: string(sample.Metric[model.LabelName("dst_namespace")]),
+		}
+		requestedNs := req.GetSelector().GetResource().GetNamespace()
+		if requestedNs != v1.NamespaceAll {
+			if requestedNs != key.srcNs && requestedNs != key.dstNs {
+				continue
 			}
 		}
+		if _, ok := edgeMap[key]; !ok {
 
-		edge := edgeTCP
-		if edgeHTTPIndex > -1 {
-			// if tcp_open_total doesn't have client_id info,
-			// use the response_total metric instead
-			if edge.ClientId == "" {
-				edge = edgesHTTP[edgeHTTPIndex]
+			clientID, err := s.getPodIdentity(string(sample.Metric[model.LabelName("pod")]), key.srcNs)
+			if err != nil {
+				return nil, err
+			}
```

Comment:
I think we should just log a warning and skip the sample in case the pod referred in a metric is no longer found, instead of erroring. Even if `tcp_open_connections` means current connections I did see errors after scaling down pods and calling `linkerd viz edges` shortly after.